### DISCUSSION TITLE
Is there any way to take control of request earlier？

### DISCUSSION THREAD

**Discussion by Hydrostic** [salvo-rs/salvo#685] [General]
I'm working on a file uploading api and I need to limit the size of body to a certain size. But I find out that either the salvo's builtin size limiter or the limiter I wrote in the handler cannot response correctly when using some clients. From the trace info from curl, it looks like that when the program runs to the handler, the client has already started to send the body, and the server's action of closing input cause the client to present a `connection closed` error. But I found nginx, when set with `client_max_body_size` option, can work correctly. I also trace the request and find out that nginx responses the header immediately after client sends its header. Is there any way to take control of request earlier and return the header after receiving the client's header?.

**Comment by chrislearn** [comment]
You can use https://docs.rs/salvo_core/0.65.2/salvo_core/http/request/fn.set_secure_max_size.html  to set request max payload size.

If your handler returns before you read the whole request payload, client will get response immediately.

**Comment by Hydrostic** [comment]
Actually even I do nothing in the handler, the error also appears. By the way I'm using the stream and the method above has no effect.

**Comment by chrislearn** [comment]
没看到你的代码 不好说

**Comment by Hydrostic** [comment]
```rust
pub async fn upload_avatar(req: &mut Request, res: &mut Response) -> AppResult<()> {
    let size_hint = req.body().size_hint().1;
    if let Some(upper) = size_hint {
        if upper > 10*1024 { 
            return Err(GlobalUserError::PayloadTooLarge.into());
        }
    } else {
        return Err(GlobalUserError::PayloadLengthRequired.into());
    }
    let mut stream = StreamReader::new(StreamBody::new(req.body_mut()));
    // 上传逻辑
    Ok(())
}
```
StreamBody 是 ReqBody 的封装，拆掉了外面的那层 `Frame`
```rust
pub struct StreamBody<'a>{
    body: Pin<&'a mut ReqBody>
}
impl<'a> StreamBody<'a>{
    pub fn new(body: &'a mut ReqBody) -> Self{
        StreamBody{body: Pin::new(body)}
    }
}
impl Stream for StreamBody<'_> {
    type Item = Result<Bytes, std::io::Error>;
    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
        self.body.as_mut().poll_next(cx).map_ok(|f| {
            f.into_data().unwrap_or(Bytes::new())
        })
    }
}

```

**Comment by chrislearn** [comment]
你要给一个最小重现的demo.zip 然后你这似乎也没必要自己整body 直接map 一下就能把frame 脱掉

**Comment by Hydrostic** [comment]
那有没有办法能先发送响应然后保持到客户端关闭连接

**Comment by chrislearn** [comment]
当然可以，可以把Body设置成stream.具体自己查阅文档

**Comment by Hydrostic** [comment]
但是好像实际上stream只是到缓冲区，并不是发送到客户端，我用wireshark抓了下包，发现貌似函数退出后才会发送tcp报文

**Comment by chrislearn** [comment]
用stream就是说你函数先返回了，应该是你没用对，这种底层设计不会有问题的。

**Comment by Hydrostic** [comment]
但是我想要的是先响应，然后同时客户端发送数据的时候服务端继续回应而不是发送reset报文，stream好像实现不了这个功能欸

**Reply by chrislearn** [reply]
可以的，然后还有 Channel 可以用。都能实现你要的功能。

https://docs.rs/salvo_core/0.65.2/salvo_core/http/response/struct.Response.html#method.channel

**Comment by Hydrostic** [comment]
channel 我也测试过了，好像也不行，是我理解有问题？
下面是最小复现样例
```rust
use std::time::Duration;

use salvo::prelude::*;
use tokio::time::sleep;

#[handler]
async fn hello2(res: &mut Response) {
    let mut chan = res.channel();
    chan.send_data("test data").await.unwrap();
    println!("now1: {}", chrono::Local::now().to_rfc3339());
    sleep(Duration::from_secs(5)).await;
    println!("now2: {}", chrono::Local::now().to_rfc3339());
}
#[tokio::main]
async fn main() {
    

    let router = Router::new().push(
        Router::with_path("/2").post(hello2)
    );
    let acceptor = TcpListener::new("127.0.0.1:2324").bind().await;
    Server::new(acceptor).serve(router).await;
}
```
测试打印的是：
```
now1: 2024-02-11T21:59:34.752292997+08:00
now2: 2024-02-11T21:59:39.754121545+08:00
```
然后我在wireshark里面抓到的包时间显示在now2之后
<img width="723" alt="image" src="https://github.com/salvo-rs/salvo/assets/45892027/176a8602-24f5-43c2-af78-2ee98dc1a82e">
我看到有一个 body closing 的特别说明， 但是我尝试 drop `chan` 好像没什么用，另一个 `abort` 方法也没有找到

**Comment by chrislearn** [comment]
你理解错了，代码写错了
